doc/developer/design/20230717_alter_set_cluster.md

```diff
@@ -0,0 +1,131 @@
+* The usual constraints apply. For sources and sinks, the cluster must either be
+  empty or only host sources and sinks. The replication factor must be 0 or 1.
+  For materialized views, the cluster must not host sources or sinks.
+* Clusters can be managed or unmanaged.
+
+When the command updates the cluster, Materialize will immediately drop the
+existing source, sink, or materialized view and recreate it in the target
+cluster.
+
+> **Warning**
+>
+> In the initial implementation, changing the cluster associated will result in
+> downtime while the object recreates. During this time, the source, sink, or
+> materialized view will not produce new data.
+>
+> We hope to change the behavior in a future release so that no downtime is
+> incurred, but this will require smarter scheduling.
+
+Users may not schedule compute objects together with sources and sinks.
+Attempting to schedule incompatible objects will result in an error like the
+following:
+
+```sql
+ALTER SOURCE src SET IN CLUSTER compute_cluster;
+ERROR: cannot alter source cluster to cluster containing indexes or materialized views
+```
+
+Users may not schedule sources or sinks on clusters with an incompatible
+replication factor. Attempting to do so will result in an error like the
+following:
+
+```sql
+ALTER SOURCE src SET IN CLUSTER cluster;
+ERROR: cannot alter source cluster to cluster with more than one replica
+```
+
+Users cannot move system objects to different clusters, and users cannot move
+user objects to system clusters.
+
+### Linked clusters
+
+Sources and sinks can have linked clusters, which are clusters owned by the
+source or sink. Altering clusters of sources and sinks will cause any previously
+linked cluster to be dropped.
+
+We do not support moving a source or sink back to a linked cluster.
+
+## Alternatives
+
+* We could wait to improve the API until we have unified clusters.
+
+* We could support moving a source or sink to a linked cluster when either a size
+  value was previously specified, or explicitly set on the `ALTER` command.
+
+### `SET CLUSTER`
+
+We rejected the syntax to handle the cluster of an object as a `WITH` parameter
+that the user could set as follows:
+
+```sql
+ALTER SOURCE src SET CLUSTER clsname;
+```
+
+Note that this is _intentionally_ lacking parens
+(`ALTER SOURCE src SET (CLUSTER = clstr);`).
+This is for consistency with the analogous PostgreSQL command
+`ALTER TABLE ... SET {SCHEMA|TABLESPACE} {schema|tblspc}`.
+The rule is that only parameters configured in the `WITH` block go inside the
+parens; parameters that have dedicated syntax, like schemas, tablespaces, and
+clusters, get dedicated `ALTER ... SET` syntax too.
+
+We could offer an alternative syntax using `IN`:
+```sql
+ALTER SOURCE src IN CLUSTER clstr;
+```
+## Open questions
+
+* Is the syntax accepted?
+
+* Should we support `ALTER SOURCE src RESET CLUSTER`? I'm not sure what the
+  semantics should be.
```

Comment:
The semantics could be to move the source back to a linked cluster. Seems like not something we want, considering that we want to phase out linked clusters in the future.